### ISSUE TITLE
-wb with vcf as -b outputs 0-length intervals

### ISSUE THREAD

**Issue by brentp** [arq5x/bedtools2#311]
this is courtesy of @jimhavrilla

given a.bed:
```
1    22    44
```

and a.vcf:
```
##fileformat=VCF4.1
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
1	32	.	A	T	0	PASS	DP=22
```

the output of

```Shell
bedtools intersect -a a.bed -b a.vcf -wb
```

is:

```
1	32	32	1	32	.	A	T	0	PASS	DP=22
```

but it should be
```
1	31	32	1	32	.	A	T	0	PASS	DP=22
```

**Comment by abjonnes** [comment]
This is actually a more general issue, relevant to regular `intersect` without `-wb` when using VCFs:

Correct behavior:

    $ cat a.bed
    1  22  44
    $ cat b.bed
    1  31  32
    $ ./bedtools intersect -a a.bed -b b.bed
    1  31  32

Compared to:

    $ cat b.vcf
    ##fileformat=VCF4.1
    #CHROM  POS  ID  REF  ALT  QUAL  FILTER  INFO
    1  32  .  A  T  0  PASS  DP=22
    $ ./bedtools intersect -a a.bed -b b.vcf
    1  32  32

I took a look, and I think the origin is the parsing of VCF entries into `VcfRecord`s at https://github.com/arq5x/bedtools2/blob/master/src/utils/FileRecordTools/Records/VcfRecord.cpp#L18, where the record's `_startPosStr` is not converted to 0-index position. When processing the intersection, the record's `_startPosStr` (still 1-indexed) is used as the start position string in the BED-formatted output (expecting 0-indexed).

**Comment by arq5x** [comment]
Thanks for pinging this issue. I just pushed a fix and new tests.

**Comment by abjonnes** [comment]
I think there are some lingering issues, e.g. if `-a` is a GFF and `-b` is a BED:

    $ cat a.gff
    1	pseudogene	exon	20	40	.	-	.	 gene_id "ENSG00000224777"; transcript_id "ENST00000424047"; exon_number "1"; gene_name "OR4F2P"; transcript_name "OR4F2P-001";
    $ cat b.bed
    1	31	32
    $ ./bedtools intersect -a a.gff -b b.bed
    1	pseudogene	exon	31	32	.	-	.	 gene_id "ENSG00000224777"; transcript_id "ENST00000424047"; exon_number "1"; gene_name "OR4F2P"; transcript_name "OR4F2P-001";

GFF coordinates are 1-indexed and inclusive, so I think the correct output should have 32 for both start and end of the GFF record. VCF doesn't suffer this issue since the `VcfRecord::print` does not actually use the `start` and `end` strings if they are supplied, but GFF does.

**Comment by arq5x** [comment]
You are right, thanks. I think that whole function needs to be refactored.

**Comment by arq5x** [comment]
I just pushed refactored (much simpler) code that addresses this based on all of my tests.
Thanks very much for catching this!
